### PR TITLE
support bin and automatically detect sampling rate

### PR DESCRIPTION
**Pull request by Nujanauss**
To test:

1. Clone or download this repository directly from GitHub.
2. Unzip the package and move the folder "mentalab-explore" into the EEGLab plugins folder.
3. Restart EEGLab and select "Import Mentalab Explore data", under File -> Import Data -> Using EEGLAB functions and plugins.

Ideally we are testing:

- BIN and CSV files open and create an EEGDATASET - try horrible names, bad files, missing files, 8 vs 4 channels etc
- Markers show, the sampling rate is correct, and the request for channel naming shows up on the plot

### CODE REVIEW

File: mentalab-explore/getSamplingRate.m [hunk 1]
```diff
@@ -0,0 +1,24 @@
+function sample_rate = getSamplingRate(exg_timestamp)
+
+i = 0;
```
Comment (tuv-mohamad-atayi) [inline mentalab-explore/getSamplingRate.m]: inconsistent indentation

**Comment by tuv-mohamad-atayi** [comment]
Converting a binary file with a sampling rate of 500hz results in this error:
```
eeglab
Some menus items hidden. Use Preference menu to show them all.
eeglab: options file is C:\Users\mh_at\eeg_options.m
Retrieving plugin versions from server...
Retreiving download statistics...
EEGLAB: adding "ICLabel" v1.3 (see >> help eegplugin_iclabel)
EEGLAB: adding "clean_rawdata" v2.4 (see >> help eegplugin_clean_rawdata) - new version 2.6 available
EEGLAB: adding "dipfit" v4.2 (see >> help eegplugin_dipfit) - new version 4.3 available
EEGLAB: adding "firfilt" v2.4 (see >> help eegplugin_firfilt)
EEGLAB: adding "mentalab-explore" v0.1 (see >> help eegplugin_explore)
EEGLAB: adding "xdfimport" v1.18 (see >> help eegplugin_xdfimport)
Warning:
A newer revision of EEGLAB (v2022.0) is available HERE.

See Release notes for more informations
You may disable this message in the File > Preferences menu.
 
eeg_checkset note: upper time limit (xmax) adjusted so (xmax-xmin)*srate+1 = number of frames
pop_importevent(): error: cannot process empty dataset
eeg_checkset note: upper time limit (xmax) adjusted so (xmax-xmin)*srate+1 = number of frames
pop_importevent(): error: cannot process empty dataset
Creating a new ALLEEG dataset 1
```
The binary file is attached here:
[DATA000.zip](https://github.com/Mentalab-hub/Mentalab-EEGLab-Plugin/files/8512658/DATA000.zip)

**Comment by tuv-mohamad-atayi** [comment]
Loaded a binary file with 250hz sampling rate and three push button markers but the event markers are not available in the EEGLAB set. 
Log:
```
eeg_checkset note: upper time limit (xmax) adjusted so (xmax-xmin)*srate+1 = number of frames
pop_importevent(): error: cannot process empty dataset
eeg_checkset note: upper time limit (xmax) adjusted so (xmax-xmin)*srate+1 = number of frames
pop_importevent(): error: cannot process empty dataset
Creating a new ALLEEG dataset 1
eeg_checkset warning: 3rd dimension size of data (1) does not match the number of epochs (0), corrected
eeg_checkset warning: number of columns in data (4587) does not match the number of points (0): corrected
eeg_checkset note: upper time limit (xmax) adjusted so (xmax-xmin)*srate+1 = number of frames
eeg_checkset warning: 3rd dimension size of data (1) does not match the number of epochs (0), corrected
eeg_checkset warning: number of columns in data (369) does not match the number of points (0): corrected
eeg_checkset note: upper time limit (xmax) adjusted so (xmax-xmin)*srate+1 = number of frames
Creating a new ALLEEG dataset 1
You are using a new version of eegplot - if you encounter problems
type "edit eegplot.m" and change FORCE_EEGPLOT_LEGACY to true
pop_eegfiltnew() - performing 1651 point bandpass filtering.
pop_eegfiltnew() - transition band width: 0.5 Hz
pop_eegfiltnew() - passband edge(s): [0.5 40] Hz
pop_eegfiltnew() - cutoff frequency(ies) (-6 dB): [0.25 40.25] Hz
pop_eegfiltnew() - filtering the data (zero-phase, non-causal)
firfilt(): |====================| 100%, ETE 00:00
Creating a new ALLEEG dataset 2
Done.
You are using a new version of eegplot - if you encounter problems
type "edit eegplot.m" and change FORCE_EEGPLOT_LEGACY to true
```
Test file:
[DATA001.zip](https://github.com/Mentalab-hub/Mentalab-EEGLab-Plugin/files/8512724/DATA001.zip)

**Review by tuv-mohamad-atayi — CHANGES_REQUESTED** [review]
Found two bugs. Added them in the command section of the PR.

**Comment by andrea-escartin** [comment]
**CSV**
Filename appear as none
Sampling rate always detected as 250
Default plot scale is not correct, only one channel or no channels are showing

**BIN**
I get the following error:
```
Error using reshape
Number of elements must not change. Use [] as one of the size inputs to automatically calculate the appropriate size for
that dimension.

Error in parseBtPacket (line 80)
            temp = reshape(temp,[nChan,nPacket]);

Error in loadbin (line 16)
        packet = parseBtPacket(fid);

Error in pop_loadfile>loadBINCSV (line 49)
        [EEG, com] = loadbin(filepath);

Error in pop_loadfile>runMain (line 19)
    [EEG, com] = loadBINCSV(filepath, ext);

Error in pop_loadfile (line 10)
    [EEG, com] = runMain([path, filename]);
 
Error while evaluating Menu Callback.
```

**Additional**
Is it possible that when the user selects Import  Mentalab data, the folder opened is the last one selected? Right now, at least in my laptop is: C:\Users\{UserName}\AppData\Roaming\MathWorks\MATLAB Add-Ons\Collections\EEGLAB

**Comment by Nujanauss** [comment]
Hi  @andrea-escartin, can you please upload the files you used in a zip so I can take a look

**Comment by andrea-escartin** [comment]
**CSV**
- Previously mentioned issues are solved solved
- The warning below is displayed:

```
eeg_checkset warning: 3rd dimension size of data (1) does not match the number of epochs (0), corrected
eeg_checkset warning: number of columns in data (3904) does not match the number of points (0): corrected
eeg_checkset note: upper time limit (xmax) adjusted so (xmax-xmin)*srate+1 = number of frames
eeg_checkset warning: 3rd dimension size of data (1) does not match the number of epochs (0), corrected
eeg_checkset warning: number of columns in data (313) does not match the number of points (0): corrected
eeg_checkset note: upper time limit (xmax) adjusted so (xmax-xmin)*srate+1 = number of frames
```

- Could not visualize the markers in Plot>Channel data (scroll)

**BIN**

- I get the error below even though the `getMarkerIdxs.m` file is in the plugin folder.

```
Warning: Unexpected end of stream. 
> In parseBtPacket (line 71)
In loadbin (line 17)
In pop_loadfile>loadBINCSV (line 52)
In pop_loadfile>runMain (line 19)
In pop_loadfile (line 10) 
Unrecognized function or variable 'getMarkerIdxs'.

Error in loadbin (line 42)
    [eeg_marker, orn_marker] = getMarkerIdxs(exg_timestamp, orn_timestamp, marker);

Error in pop_loadfile>loadBINCSV (line 52)
        [EEG, ORN, com] = loadbin(filepath);

Error in pop_loadfile>runMain (line 19)
    [EEG, ORN, com] = loadBINCSV(filepath, ext);

Error in pop_loadfile (line 10)
    [EEG, ORN, com] = runMain([path, filename]);
 
Error while evaluating Menu Callback.
```

**Pull request merged**